docs/extending_fuel.rst

```diff
@@ -0,0 +1,268 @@
+...             data[index] *= 2
+...             data = tuple(data)
+...         return data
+
+So far so good, but our transformer could be more general: what if we want to
+double ``'features'`` and ``'targets'``, or only ``'targets'``?
+
+Transformers which are applied sourcewise like our doubler should usually
+subclass from :class:`SourcewiseTransformer`. Their constructor takes an
+additional ``which_sources`` keyword argument specifying which sources to apply
+the transformer to. It's expected to be a tuple of source names. If
+``which_sources`` is ``None``, then the transformer is applied to all sources.
+Subclasses of :class:`SourcewiseTransformer` should implement a
+:meth:`transform_source_example` method and/or a :meth:`transform_source_batch`
+method, which apply on an individual source.
+
+There also exists an :class:`AgnosticSourcewiseTransformer` class for cases
+where the example and batch implementations of a sourcewise transformer are the
+same. This class requires a :meth:`transform_any_source` method to be
+implemented.
+
+>>> from fuel.transformers import AgnosticSourcewiseTransformer
+>>> 
+>>> class Doubler(AgnosticSourcewiseTransformer):
+...     def __init__(self, data_stream, **kwargs):
+...         super(Doubler, self).__init__(
+...             data_stream=data_stream,
+...             produces_examples=data_stream.produces_examples,
+...             **kwargs)
+... 
+...     def transform_any_source(self, source, _):
+...         return 2 * source
+
+Let's try this implementation on our dummy dataset.
+
+>>> target_stream = Doubler(
+...     data_stream=DataStream(
+...         dataset=dataset,
+...         iteration_scheme=batch_scheme),
+...     which_sources=('targets',))
+>>> all_stream = Doubler(
+...     data_stream=DataStream(
+...         dataset=dataset,
+...         iteration_scheme=batch_scheme),
+...     which_sources=None)
+>>> print([batch for batch in target_stream.get_epoch_iterator()])
+[(array([1, 2]), array([-2,  2])), (array([3, 4]), array([-2,  2]))]
+>>> print([batch for batch in all_stream.get_epoch_iterator()])
+[(array([2, 4]), array([-2,  2])), (array([6, 8]), array([-2,  2]))]
+
+Finally, there exists a :class:`Mapping` transformer which acts as a swiss-army
+knife transformer. In addition to a data stream, its constructor accepts a
+function which will be applied to data coming from the stream.
+
+Here's how you would implement the feature doubler using :class:`Mapping`.
+
+>>> from fuel.transformers import Mapping
+>>> 
+>>> features_index = dataset.sources.index('features')
+>>> def double(data):
+...     data = list(data)
+...     data[features_index] *= 2
+...     return tuple(data)
+>>> mapping_stream = Mapping(
+...     data_stream=DataStream(
+...         dataset=dataset, iteration_scheme=batch_scheme),
+...     mapping=double)
+>>> print([batch for batch in mapping_stream.get_epoch_iterator()])
+[(array([2, 4]), array([-1,  1])), (array([6, 8]), array([-1,  1]))]
+
```

Comment:
[Finally a comment on the content and not on the style!!] 
How should I handle the case "I want to select the source *and* the function"?